packages/eslint-plugin-template/tests/rules/button-has-type/cases.ts

```diff
@@ -0,0 +1,71 @@
+import { convertAnnotatedSourceToFailureCase } from '@angular-eslint/utils';
+import type { MessageIds } from 'packages/eslint-plugin-template/src/rules/button-has-type';
+import { INVALID_TYPE_DATA_KEY } from 'packages/eslint-plugin-template/src/rules/button-has-type';
+
+// https://www.w3.org/TR/html401/interact/forms.html#adef-type-BUTTON
+// The types are intentionally duplicated in order to de-couple the tests from the code
+const VALID_BUTTON_TYPES: readonly string[] = ['button', 'submit', 'reset'];
+const missingType: MessageIds = 'missingType';
+const invalidType: MessageIds = 'invalidType';
+
+const validCasesWithTextAttribute = VALID_BUTTON_TYPES.map(
```

Comment:
Please remove these abstractions in favour of duplication and keep everything inline in the valid cases array. Folks shouldn't have to parse test cases as they read or update them, it adds unnecessary mental overhead and harms the readability of the tests